src/serde/ser_br.rs

```diff
@@ -0,0 +1,84 @@
+// Serialization with "back-references"
+
+use std::io;
+use std::io::Cursor;
+
+use super::object_cache::{serialized_length, treehash, ObjectCache};
+use super::read_cache_lookup::ReadCacheLookup;
+use super::write_atom::write_atom;
+use crate::allocator::{NodePtr, SExp};
+use crate::node::Node;
+
+const BACK_REFERENCE: u8 = 0xfe;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+#[derive(PartialEq, Eq)]
+enum ReadOp {
+    Parse,
+    Cons,
+}
+
+// these test cases were produced by:
+
+// from chia.types.blockchain_format.program import Program
+// a = Program.to(...)
+// print(bytes(a).hex())
+// print(a.get_tree_hash().hex())
+
+pub fn node_to_stream_backrefs(node: &Node, f: &mut dyn io::Write) -> io::Result<()> {
+    let allocator = node.allocator;
+    let mut read_op_stack: Vec<ReadOp> = vec![ReadOp::Parse];
+    let mut write_stack: Vec<NodePtr> = vec![node.node];
+
+    let mut read_cache_lookup = ReadCacheLookup::new();
+
+    let mut thc = ObjectCache::new(allocator, treehash);
+    let mut slc = ObjectCache::new(allocator, serialized_length);
```

Comment:
is there a good reason not to make `ObjectCache` a non-template and have its values be both the hash and the serialized lenght?
That would seem to be both simpler code and more efficient

Reply:
I've found myself writing this pattern many times while developing clvm, so I thought it would be good to create it. I suppose I could build both the tree hash and length caches simultaneously. I am pretty skeptical it would provide much of a performance boost.

Reply:
> I think it would be good to address the clippy warnings

Done

Reply:
> as it is now, you end up computing tree hashes twice, right? That would be my main concern. In the light wallet tree hash computations shows up in the profile

If you use `ObjectCache` to calculate the hash tree of an clvm cons, it should calculate every subtree's hash exactly once. It might make sense later to provide api to pre-populate known hashes for particular subtrees. But I don't think it will be useful for compression, which is the first application here.

Reply:
I believe I was under the impression the tree-hash was the key in the cache, but it's not `NodePtr`